Added detection for untrusted domains in script content - edited existing library and added new query and tests ("the Polyfill PR")

aegilops opened the pull request:
Polyfill.io was a popular JavaScript polyflll Content Delivery Network (CDN), used to support new web browser standards on older browsers.

In February 2024 the domain was sold, and in June 2024 it was [widely publicised](https://sansec.io/research/polyfill-supply-chain-attack) that the domain had been used to serve malicious scripts. It was taken down later in that month, leaving a window where sites that used the service could have been compromised.

Subresource Integrity (SRI) would have prevented the undetected replacement of the Polyfill script code.

This CodeQL query builds on a library, `semmle/javascript/security/FunctionalityFromUntrustedSource.qll`, which was abstracted from an existing query, `javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedSource.ql`, to detect specific cases where a script from `polyfill.io` or `cdn.polyfill.io` was used without Subresource Integrity, and so the service using it was exposed to compromise.

I've modified that existing library to add `polyfill.io` and `cdn.polyfill.io` as domains that require SRI, which makes the existing query detect such uses, and I have added a `getUrl()` predicate to classes in the library so that I can filter down only to results from `polyfill.io` in a new query.

I added a specific test for `polyfill.io` use in its most common context, in an HTML script tag.

This query will work for cases in HTML or in JavaScript code, but will _not_ find cases in any other languages such as PHP or Python where server-side scripting is used to generate client-side HTML or JavaScript.

A code search across github.com hosted sites will find more results manually.

aegilops commented:
Because Polyfill.io dynamically generates its content, SRI is apparently _impossible_.

I'm going to redraft this tomorrow to account for this.

I will refactor the library I made to separate known-bad domains from those requiring SRI.

I also need to remove any references to SRI being a mitigation, and edit the Cloudflare example somehow to make it work - my current example assumes SRI will function correctly.

erik-krogh commented:
Hmmm. From reading the updates in the article: https://sansec.io/research/polyfill-supply-chain-attack it sounds like the problem has been fixed.  
The domain no longer exists (see status field in https://who.is/whois/polyfill.io), so https://polyfill.io no longer exists (you can safely click the link, it doesn't work).

aegilops commented:
I've made a few changes in the interest of getting this merged.

I've extended the query now using data extensions, so it includes more domains operated by the same CDN operator who used `polyfill.io` maliciously. Those are still active, and present a clear risk to users of them who do not use SRI.

Nevertheless, I moved it to an experimental query. I'd rather it was merged into the main queries for widespread visibility of code that is or was exposed to this risk, but will accept merging into experimental and escalating a move to the main queries later.

It also allows users to extend the query using a data extensions model pack.

I also changed the existing CDN query to allow the extension of the CDN list that requires an SRI more easily in a data extension file. It no longer requires escaping dots in a regex - you just add the domain to the next line in a string in a tuple in the YAML file.

Hope that motivates merging this.

erik-krogh commented on javascript/ql/lib/semmle/javascript/security/domains/untrusted/untrusted_domains.model.yml:
```diff
@@ -0,0 +1,14 @@
+      pack: codeql/javascript-all
+      extensible: isUntrustedDomain
+    data:
+      # new location of the polyfill.io CDN, which was used to serve malware. See: https://www.cside.dev/blog/the-polyfill-attack-explained
+      - ["polyfill.com"]
+      - ["polyfillcache.com"]
+
+      # domains operated by the same owner as polyfill.io, which was used to serve malware. See: https://www.cside.dev/blog/the-polyfill-attack-explained
+      - ["bootcdn.net"]
+      - ["bootcss.com"]
+      - ["staticfile.net"]
+      - ["staticfile.org"]
```
Nice, those are still active.  
That removes my argument for keeping the new query in the `experimental/` folder, you can move the new query to the default suite.

aegilops commented:
I've moved the query into the main set of queries, and slightly updated the `changes-notes` entry. Nothing practical changed, so 🤞 it passes cleanly in the new location

erik-krogh commented:
I did some evaluations, and they look good: ([nightly](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-16886-d71be8__nightly-old__security-extend/reports), [default](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-16886-d71be8__default__code-scanning/reports) (internal links)).  
There are some TPs that I like.  
Including a large webpage that previously used polyfill.io (but they have since changed to use the cloudfare hosted copy).  

There is a lot of documentation here, and I would like some eyes from the doc team, but the PR is good to merge after that. 
(The doc team gets pinged with the label I just added).  
I'm on PTO the next few weeks, but you can ask in #codeql-dynamic for someone else to press the merge button.  
Tell them that Erik said it's OK to merge.

felicitymay commented on javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedDomain.qhelp:
```diff
@@ -0,0 +1,98 @@
+		</p>
+
+		<p>
+			To help mitigate future risk of including a script that could be compromised, consider whether you need to
+			use a polyfill or other library at all. Modern browsers do not require a polyfill, and other popular libraries are redundant after enhancements to HTML 5.
```
Minor revision:

```suggestion
			To help mitigate the risk of including a script that could be compromised in the future, consider whether you need to
			use polyfill or another library at all. Modern browsers do not require a polyfill, and other popular libraries were made redundant by enhancements to HTML 5.
```

felicitymay commented on javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedDomain.qhelp:
```diff
@@ -0,0 +1,98 @@
+		</p>
+
+		<p>
+			If you do need a polyfill service or library, move to using a trusted CDN.
```
From the paragraph below, it seems as if you must vet CDNs yourself, so suggest:

```suggestion
			If you do need a polyfill service or library, move to using a CDN that you trust.
```

felicitymay commented on javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedDomain.qhelp:
```diff
@@ -0,0 +1,98 @@
+		    
+			A dynamic service cannot be easily used with SRI. Nevertheless,
+			it is possible to list multiple acceptable SHA hashes in the <code>integrity</code> attribute,
+			such as those for the content generated for major browers used by your users.
```
Maybe:

```suggestion
			such as hashes for the content required for the major browsers used by your users.
```

felicitymay commented on javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedDomain.qhelp:
```diff
@@ -0,0 +1,98 @@
+		<sample src="polyfill-trusted.html" />
+
+		<p>
+			If you can investigate the most used browsers by your users, you can list the hashes of the polyfills for those browsers:
```
Maybe clearer as:

```suggestion
			If you know which browsers are used by the majority of your users, you can list the hashes of the polyfills for those browsers:
```

felicitymay commented on javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedDomain.ql:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Untrusted domain used in script or other content
+ * @description Use of a script or other content from an untrusted or compromised domain
```
This doesn't follow the guidelines for query descriptions (see https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md#query-descriptions-description).

Maybe this would work?

```suggestion
 * @description Using a resource from an untrusted or compromised domain makes your code vulnerable to receiving malicious code.
```

felicitymay reviewed (approved):
A few small docs suggestions, but it's good to have help with lots of context like this 👍🏻

aegilops commented:
@felicitymay thanks so much for the suggestions!

I transferred my note about using SRI with dynamic services into the other query help too. It's the same wording in both places now.

I thought about usability and added a note on `CUSTOMIZING.md` into a separate file from the query help, and referenced its existence in the query help.

It describes how someone can extend the queries with data extensions, to add their own hostname that needs SRI, or to add their own untrusted domain.

felicitymay commented on javascript/ql/src/Security/CWE-830/CUSTOMIZING.md:
```diff
@@ -0,0 +1,43 @@
+# Functionaility from untrusted source/domain - customizations
```
Typo in the first word, but perhaps a task focused title is better?

```suggestion
# Extending the library list of untrusted sources and domains
```

felicitymay commented on javascript/ql/src/Security/CWE-830/CUSTOMIZING.md:
I suggest that we move this information into the [Customizing Library Models for JavaScript — CodeQL](https://codeql.github.com/docs/codeql-language-guides/customizing-library-models-for-javascript/) article as a practical example. Alternatively, we could create a new article in https://github.com/github/codeql/tree/main/docs/codeql/codeql-language-guides, which could be extended with future examples.

This would require converting the format to RST, but the content is fairly simple so this shouldn't be too much of an overhead and we can provide support if needed.

felicitymay commented on javascript/ql/src/Security/CWE-830/CUSTOMIZING.md:
```diff
@@ -0,0 +1,43 @@
+
+You can extend the behavior of the `js/functionality-from-untrusted-source` and `js/functionality-from-untrusted-domain` queries using [CodeQL data extensions](https://codeql.github.com/docs/codeql-language-guides/customizing-library-models-for-javascript/).
+
+This allows you to require Subresource Integrity (SRI) checks on specific content delivery network (CDN) hostnames, and add additional domains to warn on, respectively.
```
I'm not sure that "respectively" adds anything in this case.

As a non-expert, it seems to me as if step one is adding more domains and step 2 is requiring SRI check, so I'd tend to suggest swapping the sentence around.

```suggestion
This allows you to add additional domains to warn users about and to require Subresource Integrity (SRI) checks on specific content delivery network (CDN) hostnames.
```

felicitymay commented on javascript/ql/src/Security/CWE-830/CUSTOMIZING.md:
````diff
@@ -0,0 +1,43 @@
+      - ["cdn.example.com"]
+```
+
+A suitable [model pack](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/creating-and-working-with-codeql-packs#creating-a-codeql-model-pack) might be:
````
Since we include this reference in the references section at the end, we can probably safely omit it here for simplicity:

```suggestion
You would create a model pack with this information using metadata similar to that in the example below:
```

felicitymay reviewed (commented):
Thanks for adding information about how to customize the queries 💖 

This looks good to me from a docs point of view, but I'd recommend moving the new article into a location that's easier to link to and where it sits alongside related content for discoverability.

I suspect that this addition might require a 👍🏻 from the CodeQL dynamic language team 🤔

aegilops commented:
Can I propose using the current in-repo markdown approach, and coming back after RKO with a second PR to merge it into CodeQL docs, to improve things?

I'll definitely accept your suggestions when I'm back at my IDE, but I don't really want to slow down merging this to get this customisation docs part perfect.

Sound OK, or no go?

felicitymay reviewed (approved):
Thanks for the updates 👍🏻
